### PR TITLE
(#1252) Removed CollectionOf class and its direct unit test

### PR DESCRIPTION
**Pull request by fanifieiev**
This PR is for [1252](https://github.com/yegor256/cactoos/issues/1252)

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#1256](https://github.com/yegor256/cactoos/issues/1256) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#1256](https://github.com/yegor256/cactoos/issues/1256) is assigned to @iakunin[/z](https://www.0crat.com/u/iakunin), here is [why](https://www.0crat.com/footprint/C63314D6Z/93d864b5-6679-4a2f-9fab-c862f93803e0); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @paulodamaso[/z](https://www.0crat.com/u/paulodamaso) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be no monetary reward for this job

**Comment by fanifieiev** [comment]
@iakunin Please review

**Comment by fanifieiev** [comment]
@iakunin ping

File: src/main/java/org/cactoos/collection/CollectionEnvelope.java [hunk 1]
```diff
@@ -41,6 +43,14 @@
      */
     private final Collection<X> col;
 
+    /**
+     * Ctor.
+     * @param scalar Scalar to return collection
+     */
+    public CollectionEnvelope(final Scalar<Collection<X>> scalar) {
```
Comment (victornoel) [inline src/main/java/org/cactoos/collection/CollectionEnvelope.java]: @fanifieiev please don't do that, it makes no sense for the envelope to take a scala

Comment (fanifieiev) [reply src/main/java/org/cactoos/collection/CollectionEnvelope.java]: @victornoel Then several classes under `org.cactoos.collection` will suffer.

Comment (fanifieiev) [reply src/main/java/org/cactoos/collection/CollectionEnvelope.java]: @victornoel I will fix that the other way, ignore above message

Comment (fanifieiev) [reply src/main/java/org/cactoos/collection/CollectionEnvelope.java]: @victornoel  @paulodamaso I would also like to raise a discussion regarding which collections(lists, sets, etc) should be immutable and which not.
What I can see now is just a mess, some of them are read-only and some not.
What do you think ???

Comment (victornoel) [reply src/main/java/org/cactoos/collection/CollectionEnvelope.java]: @fanifieiev please see the discussions in #1242 for answers to your questions

File: src/main/java/org/cactoos/collection/Immutable.java [hunk 2]
```diff
@@ -70,7 +71,7 @@ public Immutable(final X... src) {
      * @param src Source iterable
      */
     public Immutable(final Iterable<X> src) {
-        this(new CollectionOf<>(src));
+        this(new ListOf<X>(src));
```
Comment (victornoel) [inline src/main/java/org/cactoos/collection/Immutable.java]: @fanifieiev I think here it would make more sense to just remove this constructor because it makes no sense to decorate an `Iterable` with a `Collection` decorator!

**Comment by victornoel** [comment]
@fanifieiev @paulodamaso following the discussions in #1242 I wonder if we shouldn't just remove some of the decorators here since they must be implemented in `List` or `Set` already: it's just too awkward to refer to `ListOf` inside the collection package...

File: src/main/java/org/cactoos/collection/Immutable.java (hunk 2 above)
Comment (fanifieiev) [reply src/main/java/org/cactoos/collection/Immutable.java]: @victornoel What is the sense of  this class then?

Comment (victornoel) [reply src/main/java/org/cactoos/collection/Immutable.java]: @fanifieiev it's a decorator for `Collection`, not for `Iterable`. If you want to decorate an `Iterable`, you can use the decorators in the `collection` package.

Comment (victornoel) [reply src/main/java/org/cactoos/collection/Immutable.java]: @fanifieiev also I share your view on the non sense of most of the classes in the collection package, but `Immutable` does make sense actually, it decorates a `Collection` so that it behaves as `Immutable`.

Comment (fanifieiev) [reply src/main/java/org/cactoos/collection/Immutable.java]: > @fanifieiev it's a decorator for `Collection`, not for `Iterable`. If you want to decorate an `Iterable`, you can use the decorators in the `collection` package.

@victornoel Agree, but let's take for example [org.cactoos.collection.HeadOf](https://github.com/yegor256/cactoos/blob/master/src/main/java/org/cactoos/collection/HeadOf.java). It does accept `Iterable`, but probably should accept `Collection`.
The same is for other decorators in the collection package

Comment (victornoel) [reply src/main/java/org/cactoos/collection/Immutable.java]: @fanifieiev 
> Agree, but let's take for example org.cactoos.collection.HeadOf. It does accept Iterable, but probably should accept Collection.

I believe `collection.HeadOf` is useless, because there is already `iterable.HeadOf` that does the same thing.

> I think we need to narrow the number of constructors in the decorators.

Exactly my point: if it makes sense for a decorator of collections/iterable to have an `Iterable` as argument, then it means it belongs to `iterable` package, e.g., `HeadOf`. If it doesn't make sense then it's good it is in `collection` package, e.g., `Immutable` here.

**Comment by fanifieiev** [comment]
@iakunin PING

**Comment by fanifieiev** [comment]
@0crat status

File: README.md [hunk 3]
```diff
@@ -138,10 +138,10 @@ To flatten one iterable:
 ```java
 new Joined<>(
   new Mapped<>(
-    iter -> new IterableOf<>(
-      new CollectionOf<>(iter).toArray(new Integer[]{})
-    ),
-    new IterableOf<>(new IterableOf<>(1, 2, 3, 4, 5, 6))
+      iter -> new IterableOf<>(
```
Comment (iakunin) [inline README.md]: @fanifieiev as I can see, there are 2 spaces indents in this code fragment. I suggest not to change it to 4.

File: src/main/java/org/cactoos/collection/Immutable.java [hunk 4]
```diff
@@ -56,23 +55,6 @@
      */
     private final Collection<X> col;
 
-    /**
-     * Ctor.
-     * @param src Source elements
-     */
-    @SafeVarargs
-    public Immutable(final X... src) {
```
Comment (iakunin) [inline src/main/java/org/cactoos/collection/Immutable.java]: @fanifieiev why should we remove this varargs ctor? If I'm right, we should get rid only of the `CollecitonOf`. Maybe, it's better to leave a todo for that refactoring?

File: src/main/java/org/cactoos/collection/Reversed.java [hunk 5]
```diff
@@ -54,16 +55,16 @@ public Reversed(final X... src) {
      */
     public Reversed(final Iterable<X> src) {
         super(
-            new Sticky<>(
-                new CollectionOf<>(
+            new Unchecked<>(
+                new org.cactoos.scalar.Sticky<>(
```
Comment (iakunin) [inline src/main/java/org/cactoos/collection/Reversed.java]: @fanifieiev if I'm right, this replacement is also out of scope (`CollectionOf should be deleted`). Maybe it's better to leave a todo?

File: src/main/java/org/cactoos/collection/Shuffled.java [hunk 6]
```diff
@@ -52,16 +54,16 @@ public Shuffled(final T... src) {
      */
     public Shuffled(final Iterable<T> src) {
         super(
-            new Sticky<>(
-                new CollectionOf<>(
+            new Unchecked<>(
+                new org.cactoos.scalar.Sticky<>(
```
Comment (iakunin) [inline src/main/java/org/cactoos/collection/Shuffled.java]: @fanifieiev if I'm right, this replacement is also out of scope (`CollectionOf should be deleted`). Maybe it's better to leave a todo?

File: src/main/java/org/cactoos/collection/Solid.java [hunk 7]
```diff
@@ -52,11 +56,18 @@ public Solid(final T... array) {
      */
     public Solid(final Iterable<T> src) {
         super(
-            new CollectionOf<>(
+            new Unchecked<>(
                 new org.cactoos.scalar.Solid<>(
-                    () -> new Synced<>(new Sticky<>(src))
+                    new org.cactoos.scalar.Sticky<>(
```
Comment (iakunin) [inline src/main/java/org/cactoos/collection/Solid.java]: @fanifieiev if I'm right, this replacement is also out of scope (`CollectionOf should be deleted`). Maybe it's better to leave a todo?

File: src/main/java/org/cactoos/collection/Sorted.java [hunk 8]
```diff
@@ -81,16 +80,16 @@ public Sorted(final Comparator<T> cmp, final T... src) {
      */
     public Sorted(final Comparator<T> cmp, final Iterable<T> src) {
         super(
-            new Sticky<>(
-                new CollectionOf<>(
+            new Unchecked<>(
+                new org.cactoos.scalar.Sticky<>(
```
Comment (iakunin) [inline src/main/java/org/cactoos/collection/Sorted.java]: @fanifieiev if I'm right, this replacement is also out of scope (`CollectionOf should be deleted`). Maybe it's better to leave a todo?

File: src/test/java/org/cactoos/collection/CollectionOfTest.java [hunk 9]
```diff
@@ -34,11 +34,14 @@
 
 /**
  * Test Case for {@link CollectionOf}.
- *
+ * NOTE: This test class is to be removed
```
Comment (iakunin) [inline src/test/java/org/cactoos/collection/CollectionOfTest.java]: @fanifieiev in my opinion this NOTE is duplicated by info in `@deprecated` section

File: src/test/java/org/cactoos/iterable/RangeOfTest.java [hunk 10]
```diff
@@ -55,12 +55,13 @@ public final void testIntegerRange() {
...
         MatcherAssert.assertThat(
             "Can't generate a range of fibonacci integers",
-            new CollectionOf<>(
+            new ListOf<>(
                 new RangeOf<>(
                     1,
                     100,
                     new Func<Integer, Integer>() {
                         private int last;
+
```
Comment (iakunin) [inline src/test/java/org/cactoos/iterable/RangeOfTest.java]: @fanifieiev is it really necessary to add this empty line?

**Comment by iakunin** [comment]
@victornoel special thanks for such a good review =)

File: src/main/java/org/cactoos/collection/Reversed.java (hunk 5 above)
Comment (fanifieiev) [reply src/main/java/org/cactoos/collection/Reversed.java]: @iakunin Why is out of scope?
The task is to remove `CollectionOf` right? If I remove it I have to put back sth instead right?
I can remove it, but what is supposed to be for the replacement?????

File: src/main/java/org/cactoos/collection/Shuffled.java (hunk 6 above)
Comment (fanifieiev) [reply src/main/java/org/cactoos/collection/Shuffled.java]: @iakunin What will put instead of deleted `CollectionOf`?

File: src/main/java/org/cactoos/collection/Solid.java (hunk 7 above)
Comment (fanifieiev) [reply src/main/java/org/cactoos/collection/Solid.java]: @iakunin Well, ok.

File: src/main/java/org/cactoos/collection/Sorted.java (hunk 8 above)
Comment (fanifieiev) [reply src/main/java/org/cactoos/collection/Sorted.java]: @iakunin Please see [here](https://github.com/yegor256/cactoos/pull/1256#discussion_r356633140)

File: src/test/java/org/cactoos/collection/CollectionOfTest.java (hunk 9 above)
Comment (fanifieiev) [reply src/test/java/org/cactoos/collection/CollectionOfTest.java]: @iakunin OK

File: src/test/java/org/cactoos/iterable/RangeOfTest.java (hunk 10 above)
Comment (fanifieiev) [reply src/test/java/org/cactoos/iterable/RangeOfTest.java]: @iakunin fixed

File: src/main/java/org/cactoos/collection/Immutable.java (hunk 4 above)
Comment (fanifieiev) [reply src/main/java/org/cactoos/collection/Immutable.java]: @iakunin Ok.

Comment (iakunin) [reply src/main/java/org/cactoos/collection/Immutable.java]: @fanifieiev this comment was about ctor from Iterable, but not about varargs-ctor.

Comment (victornoel) [reply src/main/java/org/cactoos/collection/Immutable.java]: @iakunin this is the same thing: this class is a decorator, it's not a way to create new collection, but to decorate them.

File: src/main/java/org/cactoos/collection/Reversed.java (hunk 5 above)
Comment (iakunin) [reply src/main/java/org/cactoos/collection/Reversed.java]: @fanifieiev yes, you should put back something. Maybe we should be able to construct `ListOf` from `Scalar`? @victornoel @paulodamaso what do you think?

It is worth noting than org.cactoos.list.ListEnvelope has [ctor from Scalar](https://github.com/yegor256/cactoos/blob/master/src/main/java/org/cactoos/list/ListEnvelope.java#L74).

File: src/main/java/org/cactoos/collection/Shuffled.java (hunk 6 above)
Comment (iakunin) [reply src/main/java/org/cactoos/collection/Shuffled.java]: @fanifieiev please see https://github.com/yegor256/cactoos/pull/1256#discussion_r356656070

File: src/main/java/org/cactoos/collection/Sorted.java (hunk 8 above)
Comment (iakunin) [reply src/main/java/org/cactoos/collection/Sorted.java]: @fanifieiev please see https://github.com/yegor256/cactoos/pull/1256#discussion_r356656070

File: src/main/java/org/cactoos/collection/Reversed.java (hunk 5 above)
Comment (victornoel) [reply src/main/java/org/cactoos/collection/Reversed.java]: @iakunin see the discussions in #1242 for more context. We are in the collection package, it makes no sense to depend on the `list` package to implement it: the collection package should contain only the strict minimum to decorate collections, but concrete implementations of collection should go in list or set packages.

Also note that ListEnvelope is going to loose its constructor with a `Scalar` in #1255.

Comment (iakunin) [reply src/main/java/org/cactoos/collection/Reversed.java]: @victornoel ok, I got it, thank you for your explanation. Closing the conversation.

File: src/main/java/org/cactoos/collection/Shuffled.java (hunk 6 above)
Comment (iakunin) [reply src/main/java/org/cactoos/collection/Shuffled.java]: @fanifieiev now (after https://github.com/yegor256/cactoos/pull/1256#discussion_r356657444) this replacement if fine for me, thank you.

File: src/main/java/org/cactoos/collection/Sorted.java (hunk 8 above)
Comment (iakunin) [reply src/main/java/org/cactoos/collection/Sorted.java]: @fanifieiev now (after https://github.com/yegor256/cactoos/pull/1256#discussion_r356657444) this replacement if fine for me, thank you.

File: src/main/java/org/cactoos/collection/Immutable.java (hunk 4 above)
Comment (iakunin) [reply src/main/java/org/cactoos/collection/Immutable.java]: @victornoel ok, got it. Let this ctor be removed.

**Comment by fanifieiev** [comment]
@iakunin please review again

**Comment by iakunin** [comment]
> @iakunin please review again

@fanifieiev thank you. There is only one comment left unresolved: https://github.com/yegor256/cactoos/pull/1256#discussion_r356619717

File: README.md (hunk 3 above)
Comment (fanifieiev) [reply README.md]: @iakunin Fixed

**Comment by iakunin** [comment]
@fanifieiev now it's done. Thank you!

@paulodamaso looks good to be merged. Can you have a look, please?

**Comment by iakunin** [comment]
@paulodamaso ping

**Comment by paulodamaso** [comment]
@rultor merge

**Comment by 0crat** [comment]
@sereshqua[/z](https://www.0crat.com/u/sereshqua) please review this job completed by @iakunin[/z](https://www.0crat.com/u/iakunin), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#1256](https://github.com/yegor256/cactoos/issues/1256) is now out of scope

**Comment by 0crat** [comment]
Code review was too long (13 days), architects (@paulodamaso) were penalized, see [§55](http://www.zerocracy.com/policy.html#55)

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso)

**Comment by sereshqua** [comment]
@0crat quality good
